overlord/devicestate/handlers_install.go

```diff
@@ -630,6 +648,171 @@ func (m *DeviceManager) doRestartSystemToRunMode(t *state.Task, _ *tomb.Tomb) er
+	if !osutil.FileExists(preseedArtifact) {
+		return false, nil
+	}
+
+	model, err := findModel(st)
+	if err != nil {
+		return false, fmt.Errorf("preseed error: cannot find model: %v", err)
+	}
+
+	preseedAs, err := readPreseedAssertion(st, model, ubuntuSeedDir, sysLabel)
+	if asserts.IsNotFound(err) {
+		return false, fmt.Errorf("preseed.tgz artifact is present but preseed assertion for brand %q, model %q and system label %q couldn't be found", model.BrandID(), model.Model(), sysLabel)
+	}
+	if err != nil {
+		return false, err
+	}
+
+	// TODO: consider a writer that feeds the file to stdin of tar and calculates the digest at the same time.
+	sha3_384, _, err := osutil.FileDigest(preseedArtifact, crypto.SHA3_384)
+	if err != nil {
+		return false, fmt.Errorf("cannot calculate preseed artifact digest: %v", err)
+	}
+
+	digest, err := base64.RawURLEncoding.DecodeString(preseedAs.ArtifactSHA3_384())
+	if err != nil {
+		return false, fmt.Errorf("cannot decode preseed artifact digest")
+	}
+	if !bytes.Equal(sha3_384, digest) {
+		return false, fmt.Errorf("invalid preseed artifact digest")
+	}
+
+	logger.Noticef("apply preseed data: %q, %q", writableDir, preseedArtifact)
+	cmd := exec.Command("tar", "--extract", "--preserve-permissions", "--preserve-order", "--gunzip", "--directory", writableDir, "-f", preseedArtifact)
+	if err := cmd.Run(); err != nil {
+		return false, err
+	}
+
+	logger.Noticef("copying snaps")
+
+	deviceSeed, err := seedOpen(dirs.SnapSeedDir, sysLabel)
+	if err != nil {
+		return false, err
+	}
+	tm := timings.New(nil)
+
+	// XXX: this is expensive, but we need snap revisions. Any way around this?
+	if err := deviceSeed.LoadAssertions(nil, nil); err != nil {
+		return false, err
+	}
+	if err := deviceSeed.LoadMeta("run", tm); err != nil {
+		return false, err
+	}
+
+	// TODO: use preseedAs for snap revisions
+	copyBlob := func(sn *seed.Snap) error {
+		sq := squashfs.New(sn.Path)
+		opts := &snap.InstallOptions{
+			MustNotCrossDevices: true,
+		}
+		pinfo := sn.PlaceInfo()
+		if pinfo.SnapRevision().Unset() {
+			pinfo = snap.MinimalPlaceInfo(pinfo.InstanceName(), snap.R(-1))
+		}
+		targetPath := filepath.Join(writableDir, pinfo.MountFile())
+		mountDir := filepath.Join(writableDir, pinfo.MountDir())
+		logger.Debugf("copying: %q to %q; mount dir=%q", sn.Path, targetPath, mountDir)
+		if _, err := sq.Install(targetPath, mountDir, opts); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	preseedSnaps := make(map[string]*asserts.PreseedSnap)
+	for _, ps := range preseedAs.Snaps() {
+		preseedSnaps[ps.Name] = ps
+	}
+
+	checkSnap := func(ssnap *seed.Snap) error {
+		ps, ok := preseedSnaps[ssnap.SnapName()]
+		if !ok {
+			return fmt.Errorf("snap %q not present in the preseed assertion", ssnap.SnapName())
+		}
+		if ps.Revision != ssnap.SideInfo.Revision.N {
+			rev := snap.Revision{N: ps.Revision}
+			return fmt.Errorf("snap %q has wrong revision %s (expected: %s)", ssnap.SnapName(), rev, ssnap.SideInfo.Revision)
```

Comment:
shouldn't rev, ssnap.SideInfo.Revision be swapped around as the former is the expected one here